### ISSUE TITLE
RethinkDB unstable on Raspberry Pi

### ISSUE THREAD

**Issue by skoehler** [rethinkdb/rethinkdb#6449]
RethinkDB is impossible to use on a Raspberry Pi 3, running Raspbian 2017-06-21 (jessie).

I'm trying to setup a cluster with 7 Raspberry Pis. The thing is that RethinkDB either hangs and takes 100% CPU (the webinterface on port 8080 is unavailable in that case) or it "crashes" with the following error:

```
Jul 10 12:02:59 raspberry2 rethinkdb[4026]: Version: rethinkdb 2.3.6 (GCC 4.9.2)
Jul 10 12:02:59 raspberry2 rethinkdb[4026]: error: Error in ./src/concurrency/exponential_backoff.hpp at line 21:
Jul 10 12:02:59 raspberry2 rethinkdb[4026]: error: Guarantee failed: [static_cast<uint64_t>(backoff_ms * fail_factor) > backoff_ms] rounding screwed it up
Jul 10 12:02:59 raspberry2 rethinkdb[4026]: error: Backtrace:
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: error: Mon Jul 10 12:02:59 2017
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: 1 [0x5a3c68]: backtrace_t::backtrace_t() at backtrace.cc:210 (discriminator 1)
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: 2 [0x5a403c]: format_backtrace(bool) at basic_string.h:275
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: 3 [0x8f1c74]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:85
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: 4 [0x68167c]: auto_reconnector_t::try_reconnect(server_id_t const&, auto_drainer_t::lock_t) at exponential_backoff.hpp:20
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: 5 [0x681914]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (auto_reconnector_t::*)(server_id_t const&, auto_drainer_t::lock_t)> (auto_reconnector_t*, server_id_t, auto_drainer_t::loc
k_t)> >::run_action() at functional:569
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: 6 [0x8d373c]: coro_t::run() at coroutines.cc:282
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: 7 [0x8d36c0]: coro_t::run() at coroutines.cc:250
Jul 10 12:03:17 raspberry2 rethinkdb[4026]: error: Exiting.
```

The error above is followed by lots and lots of "crashing while already crashed" messages which contain the "rounding screwed it up" message.  So RethinkDB does not even exit, it just continues to run.

I have traced the problem a litte, and it seems that the fail_factor is overwritten with zero at some point. Maybe backoff_ms is also overwritten, but the code will set it to 50 if it falls below that value. Anyhow, the fail_factor being zero lets the guarantee in exponential_backoff.hpp:20 to fail.

I had hoped that valgrind would shed any light on the issue, but valgrind reports all kinds of issues that may or may not be false positives.

I have compiled/installed rethinkdb on the Pi as follows:
```
apt-get install \
	protobuf-compiler libprotobuf-dev \
	libssl-dev libcurl4-openssl-dev libncurses5-dev \
	libboost-system-dev libicu-dev m4
./configure --with-system-malloc
LDFLAGS="-Wl,-s" make
make install
```

The LDFLAGS make sure the binary is strip'ed cause the unstrip'ed binary is 250 megabytes large.

The error seems to occur whenever one of the other servers in the cluster restarts. I don't think that the issue is caused by the disconnect. The memory of the exponential_backoff object is probably overwritten long before the server actually attempts a reconnect.

I cannot reproduce the issue on an x86_64 machine. I know that ARM support is experimental, but any help would be strongly appreciated.

**Comment by skoehler** [comment]
I recompiled RethinkDB with `-O1` instead of `-O3`. Now the main process crashes frequently (and systemd restarts RethinkDB).

**Comment by robertjpayne** [comment]
@skoehler do you have any older Raspberry Pi's you could test with too? Also did you try 2.3.5 rather than 2.3.6? Just curious if any difference.

I've reviewed the assembly and the ARM assembly is fine. It's probably something else that is rather buggy :(

**Comment by skoehler** [comment]
I used 2.3.5 before I switched to 2.3.6. It had the exact same problem.

**Comment by skoehler** [comment]
Now I tried rethinkdb 2.3.6 on Arch Linux for Raspberry Pi (the armv7 release, not aarch64). And I get the same error. Note that Arch Linux has substantially more recent packages.

```
Jul 09 00:28:25 raspberry3 rethinkdb[990]: Version: rethinkdb 2.3.6 (GCC 7.1.1)
Jul 09 00:28:25 raspberry3 rethinkdb[990]: error: Error in ./src/concurrency/exponential_backoff.hpp at line 21:
Jul 09 00:28:25 raspberry3 rethinkdb[990]: error: Guarantee failed: [static_cast<uint64_t>(backoff_ms * fail_factor) > backoff_ms] rounding screwed it up
Jul 09 00:28:25 raspberry3 rethinkdb[990]: error: Backtrace:
Jul 09 00:28:25 raspberry3 rethinkdb[990]: error: Sun Jul  9 00:28:25 2017
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        1 [0x2b25bc]: backtrace_t::backtrace_t() at ??:?
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        2 [0x2b29c8]: format_backtrace[abi:cxx11](bool) at ??:?
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        3 [0x2ddd60]: report_fatal_error(char const*, int, char const*, ...) at ??:?
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        4 [0x4062cc]: auto_reconnector_t::try_reconnect(server_id_t const&, auto_drainer_t::lock_t) at ??:?
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        5 [0x406530]: callable_action_instance_t<std::_Bind<void (auto_reconnector_t::*(auto_reconnector_t*, server_id_t, auto_drainer_t::lock_t))(server_id_t const&, auto_drainer_t::lock_t)> >::run_action() at ??:?
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        6 [0x81ecf8]: coro_t::run() at ??:?
Jul 09 00:28:25 raspberry3 rethinkdb[990]:        7 [0x81ecc0]: coro_t::run() at ??:?
Jul 09 00:28:25 raspberry3 rethinkdb[990]: error: Exiting.
```

I think the bug is in rethinkdb. I might add, that the network connection between the pis might be somewhat unstable - yet, rethinkdb shouldn't crash.

**Comment by noBlubb** [comment]
Hey, 
I had almost no issues building on my raspberry pi 3 running 2017-07-05-raspbian-jessie-lite:

Started from the latest tgz
```
wget https://download.rethinkdb.com/dist/rethinkdb-2.3.6.tgz
tar xf rethinkdb-2.3.6.tgz
```

And followed your instructions from above
```
apt-get install \
	protobuf-compiler libprotobuf-dev \
	libssl-dev libcurl4-openssl-dev libncurses5-dev \
	libboost-system-dev libicu-dev m4
./configure --with-system-malloc
LDFLAGS="-Wl,-s" make
```
compilation went fine, but linking would fail with an openssl related error somewhat similar to #6389. 
After I purged the `libcurl4-openssl-dev` package and reconfigured my build with `--allow-fetch` it all went fine.

```
> ./configure --allow-fetch  --with-system-malloc

* Reading arguments from 'configure.default'
* Prepending the following arguments: --fetch re2 --fetch gtest --fetch v8 --fetch jemalloc
* Detecting system configuration
Bash:                           4.3.30(1)-release
Use ccache:                     no
C++ Compiler:                   GCC 4.9.2 (/usr/bin/c++)
Host System:                    arm-linux-gnueabihf
Build System:                   Linux 4.9.35-v7+ armv7l
Cross-compiling:                no
Host Operating System:          Linux
Build Architecture:             armv7l
C++11:                          ok
Precompiled web assets:         yes
Protobuf compiler:              /usr/bin/protoc
python:                         python 2.7.9
Node.js package manager:        external/node_0.12.2
coffee:                         external/coffee-script_1.10.0
Browserify:                     external/browserify_13.1.0
bluebird:                       external/bluebird_2.9.32
web UI dependencies:            external/admin-deps_2.0.3
wget:                           /usr/bin/wget
curl:                           /usr/bin/curl
Google Test:                    external/gtest_1.7.0
termcap:                        -ltermcap
boost_system:                   -lboost_system
protobuf:                       -lprotobuf
v8 javascript engine:           external/v8_3.30.33.16-patched2
RE2:                            external/re2_20140111
z:                              -lz
crypto:                         external/openssl_1.0.1t
ssl:                            external/openssl_1.0.1t
curl:                           external/curl_7.40.0
malloc:                         system
Test protobuf:                  ok
Test boost:                     ok
Test OpenSSL:                   external/openssl_1.0.1t
Installation prefix:            /usr/local
Configuration prefix:           /usr/local/etc
Runtime data prefix:            /usr/local/var
```

```
> c++ -v

Using built-in specs.
COLLECT_GCC=c++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/arm-linux-gnueabihf/4.9/lto-wrapper
Target: arm-linux-gnueabihf
Configured with: ../src/configure -v --with-pkgversion='Raspbian 4.9.2-10' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libitm --disable-libquadmath --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-armhf/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-armhf --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-armhf --with-arch-directory=arm --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-sjlj-exceptions --with-arch=armv6 --with-fpu=vfp --with-float=hard --enable-checking=release --build=arm-linux-gnueabihf --host=arm-linux-gnueabihf --target=arm-linux-gnueabihf
Thread model: posix
gcc version 4.9.2 (Raspbian 4.9.2-10) 
```

```
> openssl version

OpenSSL 1.0.1t  3 May 2016
```

**Comment by noBlubb** [comment]
So here's my setup:
```
1 raspberry pi + 1 nextthing chip
raspberry starts first, exposes cluster port
chip starts afterwards, exposes http port, joins raspberry
restart rethinkdb on raspberry (no errors)

create 10k documents using the data explorer
restart rethinkdb on chip (no errors)

distribute a copy of the shard to raspberry
restart both after another (no errors)

split the table on both nodes
restart both after another (no errors)
```

My cluster is way smaller and the wifi connection is somewhat stable, so it's not a perfect repro. I'll leave it running for while and kill one node from time to time, maybe I can introduce some noise and disturb the connection to simulate a bad connection :)

I'm happy to test another configuration or provide further information about my build, just let me know what you need. I'll try to compile on aarch64/arm64 later/tomorrow as I can't join my current cluster that's running on x64 nodes right now, I'll update if anything unexpected occurs there.

**Comment by skoehler** [comment]
Not sure why you had trouble compiling rethink. It went smoothly, without errors, for me.

I had 5 tables, with only a few hundred documents in them. Also, I had 5 cluster nodes. Then I set the shards to 4 and the replicas to 4 for all tables. That's usually when things started to go bad. rethinkdb would either hang, crash and be restarted, or it would print the "rounding screwed it up" message above.

The latter seems to be a result of the former issue. When one rethinkdb crashed, the other ones tried to reconnect and at least one of the pis would then print the "rounding screwed it up" message after which the rethinkdb process becomes unresponsive (systemctl stop rethink would timeout and eventually forcefully kill the processes with SIGKILL).

That's all i can contribute to shed some light on the issue. Note that I was running rethinkdb on armv6 and armv7, not aarch64. I could not reproduce the issues on my PC, where I was running 7 rethinkdb instances without trouble.

Right now, I have switched to aarch64 and we switched to mongodb. Note that it will not be possible for me to easily test patches.

**Comment by regimantas** [comment]
latest 2.3.6 works on raspberry pi 2 supper good.
try recompile with clang and dont use github verion

sudo apt-get install clang
wget https://download.rethinkdb.com/dist/rethinkdb-2.3.6.tgz
tar xf rethinkdb-2.3.6.tgz
./configure --with-system-malloc --allow-fetch CXX=clang++

**Comment by skoehler** [comment]
I have tried everything, and it didn't work. I have tried Raspbian and Arch, where Arch has a much newer compiler. The problems probably don't arise unless you have a Cluster of Raspberry Pis (say 7 Pis) and enable sharding and replication (I had 4 shards and 4 replicas per shard). Furthermore, the connection was established over a Wifi ad-hoc network. The cluster would usually run fine at first and then degrade towards a crash or unresponsiveness. I don't think compiling with clang will change that.

**Comment by gabor-boros** [comment]
Since more people were able to compile on RPI with the same setup I’m closing this issue. Please let me know if I should reopen

**Comment by AlexanderNitsch** [comment]
Unfortunately the Issue seems to affect me as well.
I tried running a cluster of 3 nodes on 3 Raspberry Pi 3B+ in a LAN network.
There were only 10 very simple documents in the DB ( some usernames and passwords) and I configured it to have 1 shard and 3 replicas.

I wanted to test out the behavior if one server goes offline and comes back after some changes have been made. Killing one of the servers (By Keyboard interrupt) also caused a second server to crash with the following output:

`$ rethinkdb --bind all

Recursively removing directory /home/pi/poc-raft/rethinkdb_data/tmp
Initializing directory /home/pi/poc-raft/rethinkdb_data
Running rethinkdb 2.4.1 (CLANG 7.0.1 (tags/RELEASE_701/final))...
Running on Linux 5.10.11-v7+ armv7l
Loading data from directory /home/pi/poc-raft/rethinkdb_data
warn: Cache size does not leave much memory for server and query overhead (available memory: 744 MB).
warn: Cache size is very low and may impact performance.
Listening for intracluster connections on port 29015
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on cluster addresses: 127.0.0.1, 172.16.0.194, ::1, fe80::13db:f09f:5d07:bdc9%2
Listening on driver addresses: 127.0.0.1, 172.16.0.194, ::1, fe80::13db:f09f:5d07:bdc9%2
Listening on http addresses: 127.0.0.1, 172.16.0.194, ::1, fe80::13db:f09f:5d07:bdc9%2
Server ready, "raspberrypi_jsi" ed238421-3490-406a-9db5-0460d6bfdc0f
Connected to server "raspberrypi_srj" 1215aea1-1f4a-4ad1-8407-165ca614ea61
Connected to server "raspberrypi_dt5" 0b0b97a3-83f2-4e3e-9c06-eca04d28f2ac
Disconnected from server "raspberrypi_srj" 1215aea1-1f4a-4ad1-8407-165ca614ea61
Version: rethinkdb 2.4.1 (CLANG 7.0.1 (tags/RELEASE_701/final))
error: Error in thread 4 in ./src/concurrency/exponential_backoff.hpp at line 21:
error: Guarantee failed: [static_cast<uint64_t>(backoff_ms * fail_factor) > backoff_ms] rounding screwed it up
error: Backtrace:
Version: rethinkdb 2.4.1 (CLANG 7.0.1 (tags/RELEASE_701/final))
Crashing while already crashed. Printing error message to stderr.`

Any idea how this can be avoided?
